File: python/ampersend-sdk/src/ampersend_sdk/x402/wallet.py
from typing import Protocol

from x402_a2a import (
    PaymentPayload,
    PaymentRequirements,
)

from .types import ServerAuthorizationData


class X402Wallet(Protocol):
    def create_payment(
        self,
        requirements: PaymentRequirements,
        server_authorization: ServerAuthorizationData | None = None,
    ) -> PaymentPayload: ...

File: rust/ampersend-sdk/src/x402/wallet.rs
use async_trait::async_trait;

use super::types::{PaymentPayload, PaymentRequirements};
use crate::error::WalletError;

/// X402Wallet trait — Creates payment payloads from requirements.
///
/// An X402Wallet is responsible for creating cryptographically signed payment payloads
/// that can be submitted to sellers. Different wallet implementations support
/// different account types (EOA, smart accounts, etc.).
#[async_trait]
pub trait X402Wallet: Send + Sync {
    /// Creates a payment payload from requirements.
    async fn create_payment(
        &self,
        requirements: &PaymentRequirements,
    ) -> Result<PaymentPayload, WalletError>;
}
